### ISSUE TITLE
[Question] SatSolver: Assignment of variables

### ISSUE THREAD

**Issue by ptr1120** [google/or-tools#1013]
Hallo,

first of all thanks for the great work here. I already used the constraint solver with great success.
But with the `SatSolver` I have understanding problems with the assignment of variables:
```cs
var a = model.NewBoolVar("a");
var b = model.NewBoolVar("b");
var c = model.NewBoolVar("c");
model.AddImplication(a,b).OnlyEnforceIf(c);
```
If I now want to bind `c` to `true` in order to enforce the constraint I can find no method for that. With the `ConstraintSolver` I was able to use `SetValue(int64 v)` on a variable in order to restrict it's domain. How can I achieve similar functionality for the `SatSolver`?

Best regards,
Peter

**Comment by lperron** [comment]
C++:
model.AddEquality(c, true).

C#:
model.Add(c == true);

Please not that in your example

1) don't create c if you wish to set it to true, just add
model.AddImplication(a, b)
2) c => (a => b) <=> not(c) or (a => b) <=> not(c) or not(a) or b

model.Add BoolOr({Not(c), Not(a), b});  // in C++.

Hope I did not write something wrong. :-)
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le dim. 13 janv. 2019 à 18:49, Peter Bruch <notifications@github.com> a
écrit :

**Comment by ptr1120** [comment]
Thanks, lperron, that sounds obvious. 
But since I need to solve the model dynamically by assigning decision variables and (de-) activating constraints I see another problem with this way. 
How can I keep the state of the model? 
The ConstraintSolver had the methods `Push` and `Pop` in order to keep the initial model clean. How can I achieve similar functionality for the SatSolver?

Thanks and best regards,
Peter

**Comment by lperron** [comment]
don't

Push() and Pop() are horrible hacks.
Recreate the model and resolve.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le dim. 13 janv. 2019 à 19:09, Peter Bruch <notifications@github.com> a
écrit :

**Comment by ptr1120** [comment]
Recreating the whole model would be too much time consuming since I need to dynamically assign decision variable and (de-)activate constraints (interactive product configuration).
Is it supported to remove constraints via the `modelbuilder.MutableProto()->constraints()` or is that not  recommended because of any side effects. Or do you have any other idea?

**Comment by lperron** [comment]
Have you profiled it? 
What is the ratio between solving time and modeling time?

Have you tried with the CP-SAT solver?

**Comment by ptr1120** [comment]
Hello lperron ,

I am successfully using the ortools ConstraintSolver for product-configuration. 
But as I understood the ConstraintSolver is no longer further developed and the SatSolver has many performance impacts, that's why I want to migrate to the SatSolver.

To your question: Currently, for the ConstraintSolver the ratio between CreateMode and SolveModel (just to check whether the solution is feasible) is 10:1, so creating the model takes around 10ms and one solve costs 1-2 ms for a csp with 800 (mostly boolean) variables and 200 constraints. For one user interaction (configuration decision) in a model of such size, I need to ask the solver around 500 times  for feasibility (while manipulating the csp). That's why I think that recreating instead of reusing the csp model would affect the perfomance massivly.
Also for example, if one implements a csp explain algorithm (e.g.QuickXPlain, where you need to add/remove constraints to the csp recursivly until it is consistent) the need for manipulation the model (removing constraints) would be essential.

So basically I have the following last questions now:
* will the ConstraintSolver be further developed, or replaced by the SatSolver
* is support for manipulating the model for the SatSolver planed
* does speak anything against to remove temporarily added constraints  using `CpModelProto.Constraints.Remove()`

Best regards,
Peter

**Comment by lperron** [comment]
Hi

1) no, except for the routing part.
2) not yet.
3) you should just Clear() it instead of removing it. You can also
overwrite it after clearing it.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le mer. 16 janv. 2019 à 13:30, Peter Bruch <notifications@github.com> a
écrit :

**Comment by ptr1120** [comment]
thanks
